### PR TITLE
executor: reset the related session vars for both INSERT and REPLACE

### LINKED ISSUE #40351: ERROR 1105 (HY000): runtime error: index out of range [1] with length 1
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
CREATE TABLE `t_securities` (
  `id` bigint NOT NULL AUTO_INCREMENT COMMENT 'id',
  `security_id` varchar(8) CHARACTER SET utf8 COLLATE utf8_bin NOT NULL ,
  `market_id` smallint DEFAULT NULL ,
  `security_type` int DEFAULT NULL ,
  PRIMARY KEY (`id`) USING BTREE,
  UNIQUE KEY `unique_securityId_marketId` (`security_id`,`market_id`) 
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin  ;

INSERT INTO t_securities (security_id, market_id, security_type) values    ("1", 2,7), ("7", 1,7) ON DUPLICATE KEY UPDATE         security_type = VALUES(security_type);

replace INTO t_securities (security_id, market_id, security_type) select security_id+1, 1, security_type from t_securities  where security_id="7";

INSERT INTO t_securities (security_id, market_id, security_type) values    ("1", 2,7), ("7", 1,7) ON DUPLICATE KEY UPDATE         security_type = VALUES(security_type);
```
### 2. What did you expect to see? (Required)

success

### 3. What did you see instead (Required)

ERROR 1105 (HY000): runtime error: index out of range [1] with length 1

### 4. What is your TiDB version? (Required)
v6.1.1

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #40351 

Problem Summary:

We only reset the CurrInsertBatchExtraCols in INSERT and forgot to reset it in REPLACE. So it will cause panic in a long session.

### What is changed and how it works?

Move the reset to the func `ResetContextOfStmt `, making sure they are always reset.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the case that INSERT/REPLACE might panic in a long session connection
```

### CODE REVIEW

File: executor/writetest/write_test.go [hunk 1]
```diff
@@ -4318,3 +4318,15 @@ func TestIssueInsertPrefixIndexForNonUTF8Collation(t *testing.T) {
...
 	tk.MustGetErrCode("insert into t3 select 'abc d'", 1062)
 }
+
+func TestMutipleReplaceAndInsertInOneSession(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("create table t_securities(id bigint not null auto_increment primary key, security_id varchar(8), market_id smallint, security_type int, unique key uu(security_id, market_id))")
+	tk.MustExec(`insert into t_securities (security_id, market_id, security_type) values ("1", 2, 7), ("7", 1, 7) ON DUPLICATE KEY UPDATE security_type = VALUES(security_type)`)
+	tk.MustExec(`replace into t_securities (security_id, market_id, security_type) select security_id+1, 1, security_type from t_securities  where security_id="7";`)
```
Comment (bestwoody) [inline executor/writetest/write_test.go:4328]: how about add another TestCase with `"insert into ... from .. where.."` like `"replace into ... from .. where.."`

**Review by bestwoody — APPROVED** [review]
lgtm

**Comment by bestwoody** [comment]
my LGTM wasn't count. "The bot only counts approvals from reviewers and higher roles in [list](https://prow.tidb.net/tichi/repos/pingcap/tidb/pulls/40354/owners)"

**Review by wshwsh12 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

File: executor/executor.go [hunk 2]
```diff
@@ -2186,6 +2186,8 @@ func ResetContextOfStmt(ctx sessionctx.Context, s ast.StmtNode) (err error) {
 	vars.PrevFoundInBinding = vars.FoundInBinding
 	vars.FoundInBinding = false
 	vars.DurationWaitTS = 0
+	vars.CurrInsertBatchExtraCols = nil
```
Comment (XuHuaiyu) [inline executor/executor.go:2189]: Why set to nil instead of resetting the old one?

Comment (winoros) [reply executor/executor.go:2189]: It cannot guarantee that the slice header here is not nil.

**Pull request merged**